File: 2025/ocaml/day04.ml
let loop_3x3 fn init =
  let loop3 = List.init 3 Fun.id in
  let accumulator i acc j = fn acc j i in
  loop3
  |> List.fold_left
       (fun accum i -> List.fold_left (accumulator i) accum loop3)
       init

let is_invalid_roll mat r c =
  let nrow = Array.length mat in
  let ncol = Array.length mat.(0) in
  Int.min r c < 0 || r >= nrow || c >= ncol || mat.(r).(c) != '@'

let is_accessible mat r c =
  if mat.(r).(c) != '@' then false
  else
    let accumulator roll_cnt i j =
      let rr, cc = (r + i - 1, c + j - 1) in

      if (r, c) = (rr, cc) then roll_cnt
      else if is_invalid_roll mat rr cc then roll_cnt
      else roll_cnt + 1
    in

    loop_3x3 accumulator 0 < 4

let part_one input =
  let mat =
    input |> String.split_on_char '\n'
    |> List.map (fun l -> l |> String.to_seq |> Array.of_seq)
    |> Array.of_list
  in

  let nrow = Array.length mat in
  let ncol = Array.length mat.(0) in
  let loop_rows = List.init nrow Fun.id in
  let loop_cols = List.init ncol Fun.id in

  loop_rows
  |> List.fold_left
       (fun accessible_cnt row ->
         loop_cols
         |> List.fold_left
              (fun accessible_cnt col ->
                accessible_cnt + if is_accessible mat row col then 1 else 0)
              accessible_cnt)
       0

let remove_and_notify_neighbor mat work_queue r c =
  Array.set mat.(r) c '.';

  let accumulator () i j =
    let rr, cc = (r + i - 1, c + j - 1) in
    if is_invalid_roll mat rr cc then () else Queue.add (rr, cc) work_queue
  in

  loop_3x3 accumulator ()

let part_two input =
  let mat =
    input |> String.split_on_char '\n'
    |> List.map (fun l -> l |> String.to_seq |> Array.of_seq)
    |> Array.of_list
  in

  let nrow = Array.length mat in
  let ncol = Array.length mat.(0) in
  let loop_rows = List.init nrow Fun.id in
  let loop_cols = List.init ncol Fun.id in

  let remove_list =
    loop_rows
    |> List.fold_left
         (fun remove_list row ->
           loop_cols
           |> List.fold_left
                (fun remove_list col ->
                  if is_accessible mat row col then (row, col) :: remove_list
                  else remove_list)
                remove_list)
         []
  in

  let work_queue = remove_list |> List.to_seq |> Queue.of_seq in

  let rec loop remove_cnt =
    match Queue.take_opt work_queue with
    | None -> remove_cnt
    | Some (row, col) ->
        if not (is_accessible mat row col) then loop remove_cnt
        else (
          remove_and_notify_neighbor mat work_queue row col;
          loop (remove_cnt + 1))
  in

  loop 0

let () =
  let part = Aoc.Util.parse_args () in
  let input = Aoc.Util.read_input_from_stdin () in

  let result =
    match part with PartOne -> part_one input | PartTwo -> part_two input
  in

  Printf.printf "Result: %d\n" result

File: 2025/zig/day04.zig
const std = @import("std");

const util = @import("util");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const alloc = gpa.allocator();
    defer _ = gpa.deinit();

    const args = try util.parseArgs();
    const input = try util.readInputFromStdin(alloc);
    defer alloc.free(input);

    const result = switch (args.part) {
        1 => try part1(alloc, input),
        2 => try part2(alloc, input),
        else => @panic("Illegal part"),
    };

    std.debug.print("Result: {}\n", .{result});
}

fn part1(_: std.mem.Allocator, input: []u8) !i64 {
    var rows = std.mem.splitScalar(u8, input, '\n');
    var nrow: usize = 0;
    var ncol: usize = 0;
    while (rows.next()) |row| : (nrow += 1) {
        if (row.len == 0) {
            break;
        }
        ncol = row.len;
    }

    var n_accessible: i64 = 0;
    for (0..nrow) |i| {
        for (0..ncol) |j| {
            if (isAccessible(input, nrow, ncol, i, j)) {
                n_accessible += 1;
            }
        }
    }

    return n_accessible;
}

fn isAccessible(mat: []u8, nrow: usize, ncol: usize, r: usize, c: usize) bool {
    const row_len = ncol + 1;
    if (mat[r * row_len + c] != '@') {
        return false;
    }
    var r_start: usize = 0;
    if (r > 0) {
        r_start = r - 1;
    }
    var r_end = nrow;
    if (r + 2 <= nrow) {
        r_end = r + 2;
    }
    var c_start: usize = 0;
    if (c > 0) {
        c_start = c - 1;
    }
    var c_end = ncol;
    if (c + 2 <= ncol) {
        c_end = c + 2;
    }

    var roll_cnt: usize = 0;
    for (r_start..r_end) |rr| {
        for (c_start..c_end) |cc| {
            if (r == rr and c == cc) {
                continue;
            }
            if (mat[rr * row_len + cc] == '@') {
                roll_cnt += 1;
            }
        }
    }

    return roll_cnt < 4;
}

fn part2(alloc: std.mem.Allocator, input: []u8) !i64 {
    var rows = std.mem.splitScalar(u8, input, '\n');
    var nrow: usize = 0;
    var ncol: usize = 0;
    while (rows.next()) |row| : (nrow += 1) {
        if (row.len == 0) {
            break;
        }
        ncol = row.len;
    }

    // Holds pairs of (row, col) tuples
    var work_queue = std.ArrayList(usize).empty;
    defer work_queue.deinit(alloc);
    try work_queue.ensureTotalCapacity(alloc, nrow * ncol * 2);

    for (0..nrow) |i| {
        for (0..ncol) |j| {
            if (isAccessible(input, nrow, ncol, i, j)) {
                try work_queue.append(alloc, i);
                try work_queue.append(alloc, j);
            }
        }
    }

    var n_accessible: i64 = 0;
    while (work_queue.items.len > 1) {
        const c = work_queue.pop().?;
        const r = work_queue.pop().?;
        if (isAccessible(input, nrow, ncol, r, c)) {
            try removeSelfAndNotifyNeighbor(alloc, &work_queue, input, nrow, ncol, r, c);
            n_accessible += 1;
        }
    }
    return n_accessible;
}

fn removeSelfAndNotifyNeighbor(alloc: std.mem.Allocator, work_queue: *std.ArrayList(usize), mat: []u8, nrow: usize, ncol: usize, r: usize, c: usize) !void {
    const row_len = ncol + 1;

    mat[r * row_len + c] = '.';

    var r_start: usize = 0;
    if (r > 0) {
        r_start = r - 1;
    }
    var r_end = nrow;
    if (r + 2 <= nrow) {
        r_end = r + 2;
    }
    var c_start: usize = 0;
    if (c > 0) {
        c_start = c - 1;
    }
    var c_end = ncol;
    if (c + 2 <= ncol) {
        c_end = c + 2;
    }

    for (r_start..r_end) |rr| {
        for (c_start..c_end) |cc| {
            if (mat[rr * row_len + cc] == '@') {
                try work_queue.append(alloc, rr);
                try work_queue.append(alloc, cc);
            }
        }
    }
}
